### PR TITLE
[Enhancement] support exchanging >=2GB message via http rpc

### PR DESCRIPTION
**Pull request by fzhedu**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #https://github.com/StarRocks/starrocks/issues/15316

## Problem Summary(Required) ：

if sending meg's size > `rpc_http_min_size`, it will use HTTP RPC, otherwise using the std baidu_rpc.

HTTP RPC can directly send a request proto with requiring the attachment should be empty. However when the attachment is not empty, the request proto should be put into attachment explicitly.

## Results

before this PR, RPC error when the RPC msg's size > 2GB
after this PR, the large msg transmit successfully via HTTP RPC.

```
// force the standard RPC failed due to the large size of the attachment.

mysql> select host_name() as h, length( group_concat(concat(l_shipinstruct,l_comment)))/1024/1024/1024 as siz from lineitem group by h;
ERROR 1064 (HY000): transmit chunk rpc failed:29ea49f3-c1a8-11ed-896b-00163e132524


// force RPC via HTTP, works fine.

mysql> select host_name() as h, length( group_concat(concat(l_shipinstruct,l_comment)))/1024/1024/1024 as siz from lineitem group by h;
+-------------------------+----------------------+
| h                       | siz                  |
+-------------------------+----------------------+
| iZ8vbgtsjtof0smhp0fd7lZ | -0.45599549543112516 |
| iZ8vbgtsjtof0smhp0fd7kZ |  -0.4556004209443927 |
| iZ8vbgtsjtof0smhp0fd7mZ | -0.45604985300451517 |
+-------------------------+----------------------+
3 rows in set (53.45 sec)
```

test on tpcds-100g,[ the performance seems no differ if using http rpc.](https://starrocks.feishu.cn/sheets/shtcnRkmYLglImO2lTHABonJlx7)
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|2|50.00%|[1899]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|2|50.00%|[1899]|

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 1]
```diff
@@ -370,20 +371,45 @@ Status SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id, const std::fun
...
         closure->cntl.set_timeout_ms(_brpc_timeout_ms);
-        closure->cntl.request_attachment().append(request.attachment);
 
-        if (bthread_self()) {
-            request.brpc_stub->transmit_chunk(&closure->cntl, request.params.get(), &closure->result, closure);
-        } else {
+        if (!bthread_self()) {
             // When the driver worker thread sends request and creates the protobuf request,
             // also use process_mem_tracker to record the memory of the protobuf request.
             SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(nullptr);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:378]: `SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER` is only effective in the block of `if (!bthread_self()) {...}`, so `_send_rpc` **MUST** be invoked in this block `if (!bthread_self())`.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 2 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|2|50.00%|[1899]|

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:382]: fixed.

File: be/src/util/brpc_stub_cache.h [hunk 2]
```diff
@@ -84,6 +84,37 @@ class BrpcStubCache {
         return stub;
     }
 
+    // rarely used, so create as needed
+    static std::shared_ptr<doris::PBackendService_Stub> create_http_stub(const TNetworkAddress& taddr) {
```
Comment (silverbullet233) [inline be/src/util/brpc_stub_cache.h:88]: is it better to use unique_ptr?

File: be/src/runtime/runtime_state.h [hunk 3]
```diff
@@ -77,6 +77,8 @@ namespace pipeline {
 class QueryContext;
 }
 
+constexpr int64_t kRpcHttpMinSize = ((1L << 31) - (1L << 20));
```
Comment (silverbullet233) [inline be/src/runtime/runtime_state.h:80]: why not use (1<<31) ? substract (1<<20) for what?

Comment (fzhedu) [reply be/src/runtime/runtime_state.h:80]: avoid the case the serialized size may be larger than pre-computed.

File: be/src/util/brpc_stub_cache.h [hunk 4]
```diff
@@ -84,6 +84,37 @@ class BrpcStubCache {
...
+            }
+        }
+        if (str2endpoint(realhost.c_str(), taddr.port, &endpoint)) {
+            LOG(WARNING) << "unknown endpoint, host = " << taddr.hostname;
+            return nullptr;
+        }
+        // create
+        brpc::ChannelOptions options;
+        options.connect_timeout_ms = 3000;
+        options.protocol = "http";
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply be/src/util/brpc_stub_cache.h:106]: grpc also has 2GB limit. the offical suggestion is using http.

File: gensrc/proto/doris_internal_service.proto [hunk 5]
```diff
@@ -59,6 +59,7 @@ service PBackendService {
 
     // Transmit vectorized data between backends
     rpc transmit_chunk(starrocks.PTransmitChunkParams) returns (starrocks.PTransmitChunkResult);
+    rpc transmit_chunk_via_http(starrocks.PHttpRequest) returns (starrocks.PTransmitChunkResult);
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply gensrc/proto/doris_internal_service.proto:62]: we should add another interface, as it's logic differs from transmit_chunk(), as it should serialize proto objects into attachment first.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 6]
```diff
@@ -331,6 +331,7 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
 
     public static final String QUERY_CACHE_AGG_CARDINALITY_LIMIT = "query_cache_agg_cardinality_limit";
     public static final String TRANSMISSION_ENCODE_LEVEL = "transmission_encode_level";
+    public static final String RPC_HTTP_MIN_SIZE = "rpc_http_min_size";
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:334]: It should be upper_size instead of lower_size?
`PRC_PACKET_THRESHOLD_SIZE`

Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:334]: if the real packet's size > `rpc_http_min_size`, the rpc should be via http. it's also the max size of the std rpc.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

File: gensrc/proto/doris_internal_service.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzhedu) [reply gensrc/proto/doris_internal_service.proto:62]: ok ,delete it.

File: be/src/service/internal_service.cpp [hunk 7]
```diff
@@ -190,6 +190,50 @@ void PInternalServiceImplBase<T>::_transmit_chunk(google::protobuf::RpcControlle
...
+            if (attachment_size != iobuf.size()) {
+                Status st = Status::InternalError(
+                        fmt::format("{} != {} during deserialization via http", attachment_size, iobuf.size()));
+                return st;
+            }
+            return Status::OK();
+        };
+        Status st;
+        // catch the exceptions, like the std::bad_alloc caused by deserialization.
+        TRY_CATCH_ALL(st, get_params());
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply be/src/service/internal_service.cpp]: deleted

File: gensrc/proto/doris_internal_service.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzhedu) [reply gensrc/proto/doris_internal_service.proto:62]: cann't delete it as
```
// NOTE(zc): If you want to add new method here,
// you must add same method to 'doris_internal_service.proto'.
service PInternalService {
}
```

File: gensrc/proto/internal_service.proto [hunk 8]
```diff
@@ -63,6 +63,9 @@ message PTransmitDataParams {
     optional PQueryStatistics query_statistics = 8;
 };
 
+message PHttpRequest {
```
Comment (chaoyli) [inline gensrc/proto/internal_service.proto:66]: What's the usage of this empty protobuf message?

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 9]
```diff
@@ -370,20 +371,54 @@ Status SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id, const std::fun
...
+            LOG(INFO) << "issue a http rpc, attachment's size = " << attachment_size
+                      << " , total size = " << closure->cntl.request_attachment().size();
+        }
+        if (UNLIKELY(expected_iobuf_size != closure->cntl.request_attachment().size())) {
+            LOG(WARNING) << "http rpc expected iobuf size " << expected_iobuf_size << " != "
+                         << " real iobuf size " << closure->cntl.request_attachment().size();
+        }
+        closure->cntl.http_request().set_content_type("application/proto");
+        // create http_stub as needed
+        auto http_stub = BrpcStubCache::create_http_stub(request.brpc_addr);
```
Comment (chaoyli) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:415]: need to check whether the http_stub is a nullptr or not.

File: gensrc/proto/internal_service.proto (hunk 8 above)
Comment (fzhedu) [reply gensrc/proto/internal_service.proto:66]: just a placeholder but necessary.

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 9 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp]: updated by return status

**Review by chaoyli — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by fzhedu** [comment]
@Mergify backport-2.5

**Comment by fzhedu** [comment]
@mergify backport-3.0

**Comment by fzhedu** [comment]
@mergify backport branch-2.5

**Comment by fzhedu** [comment]
@mergify backport branch-3.0

**Comment by Karssadin** [comment]
i see the summary is that brpc not support 2G size request, but i see this content in brpc doc: `brpc removes the restriction from protobuf and controls the limit by -max_body_size solely: as long as the flag is large enough, messages will not be rejected and error logs will not be printed. This feature works for all versions of protobuf.`

but there was a limit in config.h 'CONF_Int64(brpc_max_body_size, "2147483648");', it set the brpc_limit to 2G, so cannot send 2G request.

i wanna to know: why use http to send when reach 2G rather than set the `brpc_max_body_size`  bigger

**Comment by stdpain** [comment]
> i see the summary is that brpc not support 2G size request, but i see this content in brpc doc: `brpc removes the restriction from protobuf and controls the limit by -max_body_size solely: as long as the flag is large enough, messages will not be rejected and error logs will not be printed. This feature works for all versions of protobuf.`
> 
> but there was a limit in config.h 'CONF_Int64(brpc_max_body_size, "2147483648");', it set the brpc_limit to 2G, so cannot send 2G request.
> 
> i wanna to know: why use http to send when reach 2G rather than set the `brpc_max_body_size` bigger

see https://github.com/apache/brpc/blob/master/src/brpc/policy/baidu_rpc_meta.proto 
```
message RpcMeta {
    optional RpcRequestMeta request = 1;
    optional RpcResponseMeta response = 2;
    optional int32 compress_type = 3;
    optional int64 correlation_id = 4;
    optional int32 attachment_size = 5;
    optional ChunkInfo chunk_info = 6;
    optional bytes authentication_data = 7;
    optional StreamSettings stream_settings = 8;
    map<string, string> user_fields = 9;
    optional ContentType content_type = 10;
    optional int32 checksum_type = 11;
    optional bytes checksum_value = 12;
}
```

In the rpc protocol, attachment_size is i32.

**Comment by Karssadin** [comment]
> In the rpc protocol, attachment_size is i32.

Thanks for u reply, i got it.

**Pull request merged**